ping-payload-extensions/extensions/type-3.md

```diff
@@ -0,0 +1,34 @@
+# Get Client info
+
+This payload MUST be available from all sub-networks. This payload allows you to figure out peer's client info. Which will be useful for censuses.
```

Comment:
I think this shouldn't be a `MUST` .

Either that or leave it as MUST but allow the client to decide on what information to put there (Including empty string).

Reply:
I tend to agree.  We can't force clients to divulge this information and there are going to be situations where it's in their best interest to withold it.  Better to allow them to not support it than to require them to give us garbage data.

Reply:
I wrote a response here https://github.com/ethereum/portal-network-specs/pull/348#discussion_r1851993424

I am fine with an empty string being a valid response for privacy focused users, but as I wrote in my post above I think it makes sense for "Get Client Info" to be a standard extension, as it will be supported likely by 99.9% of users, and other reasons I wrote in the link above.

Reply:
It seems redundant to have the same information exposed on multiple subnetworks. Not a big issue but would lead to duplicated code in implementations perhaps. Why not just add this an a JSON RPC endpoint? That why it can be exposed if needed or hidden for privacy reasons if desirable.

Reply:
> It seems redundant to have the same information exposed on multiple subnetworks. Not a big issue but would lead to duplicated code in implementations perhaps. Why not just add this an a JSON RPC endpoint? That why it can be exposed if needed or hidden for privacy reasons if desirable.

I don't understand this comment. I don't understand how a P2P protocol change relates to the user facing JSON RPC. A privacy option if wanted would probably be enabled through a commandline param on startup, doesn't make much sense to make it togglable through a json-rpc

Reply:
> I don't understand this comment. I don't understand how a P2P protocol change relates to the user facing JSON RPC. 

Ignoring my comment about JSON-RPC for a moment, what I meant was that returning client info for each subnetwork seems like duplicate/redundant information. For example if I run a Fluffy node and start subnetworks, state, history and beacon, the client info will be the same for all subnetworks so perhaps this information shouldn't be returned at the sub-network level. This is not a big deal, just pointing it out, that is all.

Reply:
And that is why I suggested JSON-RPC as an alternative way for returning this information.

Reply:
> And that is why I suggested JSON-RPC as an alternative way for returning this information.

A JSON-RPC wouldn't change anything this PR proposes though. It wouldn't change `the client info will be the same for all subnetworks so perhaps this information shouldn't be returned at the sub-network level` so I am not sure I understand what is being stated here.

Reply:
`portal_*Ping` are JSON-RPC endpoints so why couldn't we return this information in a different JSON-RPC endpoint for doing census and then remove the client info ping message from the protocol? This would simplify the implementation. Is there another reason for using the client info details at the subnetwork level that I'm missing?

Reply:
> `portal_*Ping` are JSON-RPC endpoints so why couldn't we return this information in a different JSON-RPC endpoint for doing census and then remove the client info ping message from the protocol? This would simplify the implementation. Is there another reason for using the client info details at the subnetwork level that I'm missing?

> `portal_*Ping`

 is done at the subnetwork level. 

> return this information in a different JSON-RPC endpoint for doing census and then remove the client info ping message from the protocol

Adding a new JSON-RPC endpoint would still require having a ping/pong type message over the p2p protocol. A `JSON-RPC endpoint` only would change the interface for the user, it wouldn't remove the P2P ping/pong message.

>  Is there another reason for using the client info details at the subnetwork level that I'm missing?

How would you propose we do this at the Discv5 level? As that is the only level above the subnetwork level.

Reply:
> 
> is done at the subnetwork level.

Yes, I understand that.

> Adding a new JSON-RPC endpoint would still require having a ping/pong type message over the p2p protocol. A `JSON-RPC endpoint` only would change the interface for the user, it wouldn't remove the P2P ping/pong message.

Perhaps I miss understanding something here. If the client info message is mainly useful for audit via glados which is using JSON-RPC then why would you need client info messages of the p2p protocol? Couldn't Glados connect to each node via JSON-RPC? Or does Glados only connect to a single node via JSON-RPC? That might be the part I missed now that I think about it.

Reply:
> > Adding a new JSON-RPC endpoint would still require having a ping/pong type message over the p2p protocol. A `JSON-RPC endpoint` only would change the interface for the user, it wouldn't remove the P2P ping/pong message.
> 
> Perhaps I miss understand something here. If the client info message is mainly useful for audit via glados which is using JSON-RPC then why would you need client info messages in the p2p protocol? Couldn't Glados connect to each node via JSON-RPC? Or does Glados only connect to a single node via JSON-RPC? That might be the part I missed now that I think about it.

Glados can only use the JSON-RPC for the Portal client it is running itself. JSON-RPC's are private interfaces to a Portal client, sometimes altruistic people will make them public for others to use. 

From glados perspective this is what would happen
- node1: call `portal_*Ping` to ENR X through the JSON-RPC
- node1: ping message is sent over the subnetwork level to the PEER
- node2: PEER sends pong back, with requested information
- node1: receives pong and returns it from the P2P subnetwork level, to the JSON-RPC handling using a callback 
- node1: JSON-RPC returns the pong response received from the PEER to the USER

> Glados only connect to a single node via JSON-RPC

So in short Glados only uses 1 Portal client's JSON-RPC. Then uses JSON-RPC requests which make Portal Wire Protocol messages at the subnetwork level.

Glados can't call another Portal clients JSON-RPC, but it can send P2P requests at the subnetwork level to other Portal clients. The JSON-RPC is the interface the user in this case uses to make `P2P requests at the subnetwork level to other Portal clients`.

So the idea is you can get the `client info` for clients/peers you don't run and hence don't have JSON-RPC access to, but you can make P2P requests to get the info through your portal clients json-rpc.